test: add tests for curated biomodels for new and old math types

sankha555 opened the pull request:
* [x] fix #199 
* [x] added [curated biomodels](https://www.ebi.ac.uk/biomodels/search?offset=0&numResults=1000&sort=relevance-desc&query=*%3A*+AND+curationstatus%3A%22Manually+curated%22&domain=biomodels) to test models for report creations and ran pytests on them.
* [x] remove uncompressed files from history

sankha555 commented:
@matthiaskoenig I have added all (984) models to the ```test/data/models/biomodels_curated/``` directory downloaded from the URL. Below are the test statistics:

Total Tests : 3936
Passed : 3858
Failed : 78
Pass Percentage : 98.01%

Most of the models that have failed are due to the same error as the previous ones, unable to parse piecewise lambda functions. I had run the tests on the latest release of the repository (0.4.9). The detailed error logs for each model are:


--------------------------------------------------------------------------------------------------------------------
Model BIOMD0000000025.xml in math_type "latex"
```
expr = eval(
>           code, global_dict, local_dict)  # take local objects in preference
             File "<string>", line 1
            Piecewise ((Integer (0 ),(Function ('delay' )(dClk tau1 )-Function ('delay' )(Per tau1 ))<Integer (0 )),(Function ('delay' )(dClk tau1 )-Function ('delay' )(Per tau1 ),True ))
                                                                 ^
       SyntaxError: invalid syntax

../../sb/lib/python3.7/site-packages/sympy/parsing/sympy_parser.py:903: SyntaxError
```

Model BIOMD0000000128.xml in math_type "latex"
```
E           sympy.core.sympify.SympifyError: Sympify of expression 'could not parse 'lambda * (ninf - n) / taun'' failed, because of exception being raised:
E           TokenError: Starred arguments in lambda not supported

../../sb/lib/python3.7/site-packages/sympy/core/sympify.py:482: SympifyError
```

Model BIOMD0000000138.xml in math_type "latex"
```
E                       tokenize.TokenError: Starred arguments in lambda not supported

../../sb/lib/python3.7/site-packages/sympy/parsing/sympy_parser.py:611: TokenError
```

Model BIOMD0000000237.xml in math_type "latex"
```
E           sympy.core.sympify.SympifyError: Sympify of expression 'could not parse 'alphastim * Piecewise((alphaA * (1 - exp(-(time - alphat) / alphas)), (time >= alphat) & (time <= alphae)),(Piecewise((alphaA * exp(-(time - alphat) / alphas) time >= alphae 0, True)), True))'' failed, because of exception being raised:
E           SyntaxError: invalid syntax (<string>, line 1)

../../sb/lib/python3.7/site-packages/sympy/core/sympify.py:482: SympifyError
```

Model BIOMD0000000248.xml in math_type "latex"
```
E           sympy.core.sympify.SympifyError: Sympify of expression 'could not parse 'Piecewise((VmaxM, exercise_level == 1),(Piecewise((VmaxH exercise_level == 2 VmaxVH, True)), True))'' failed, because of exception being raised:
E           SyntaxError: invalid syntax (<string>, line 1)

../../sb/lib/python3.7/site-packages/sympy/core/sympify.py:482: SympifyError
```

Model BIOMD0000000255.xml in math_type "latex"
```
 sympy.core.sympify.SympifyError: Sympify of expression 'could not parse 'Piecewise((v_start, t < t_start),(Piecewise((v_end t > t_end (v_start + v_end) / 2 + ((v_end - v_start) / 2) * sin((3.14 / 2) * (2 * t - t_end - t_start) / (t_end - t_start)), True)), True))'' failed, because of exception being raised:
E           SyntaxError: invalid syntax (<string>, line 1)

../../sb/lib/python3.7/site-packages/sympy/core/sympify.py:482: SympifyError
```

Similar errors were raised for the following models too:
BIOMD0000000273.xml
BIOMD0000000326.xml
BIOMD0000000373.xml
BIOMD0000000376.xml
BIOMD0000000412.xml
BIOMD0000000413.xml
BIOMD0000000429.xml
BIOMD0000000445.xml
BIOMD0000000446.xml
BIOMD0000000450.xml
BIOMD0000000532.xml
BIOMD0000000547.xml
BIOMD0000000564.xml
BIOMD0000000568.xml
BIOMD0000000577.xml
BIOMD0000000578.xml
BIOMD0000000580.xml
BIOMD0000000582.xml
BIOMD0000000589.xml
BIOMD0000000597.xml
BIOMD0000000598.xml
BIOMD0000000603.xml
BIOMD0000000604.xml
BIOMD0000000605.xml
BIOMD0000000606.xml
BIOMD0000000607.xml
BIOMD0000000610.xml
BIOMD0000000620.xml
BIOMD0000000621.xml
BIOMD0000000637.xml
BIOMD0000000638.xml
BIOMD0000000658.xml
BIOMD0000000660.xml
BIOMD0000000664.xml
BIOMD0000000669.xml
BIOMD0000000672.xml
BIOMD0000000673.xml
BIOMD0000000674.xml
BIOMD0000000675.xml
BIOMD0000000683.xml
BIOMD0000000685.xml
BIOMD0000000686.xml
BIOMD0000000687.xml
BIOMD0000000688.xml
BIOMD0000000695.xml
BIOMD0000000698.xml
BIOMD0000000700.xml
BIOMD0000000726.xml
BIOMD0000000731.xml
BIOMD0000000739.xml
BIOMD0000000757.xml
BIOMD0000000759.xml
BIOMD0000000775.xml
BIOMD0000000784.xml
BIOMD0000000801.xml
BIOMD0000000802.xml
BIOMD0000000814.xml
BIOMD0000000819.xml
BIOMD0000000840.xml
BIOMD0000000865.xml
BIOMD0000000875.xml
BIOMD0000000876.xml
BIOMD0000000879.xml
BIOMD0000000887.xml
BIOMD0000000887.xml
BIOMD0000000948.xml



--------------------------------------------------------------------------------------------------------------------
In model BIOMD0000000955.xml in math type "latex"
```
E           sympy.core.sympify.SympifyError: Sympify of expression 'could not parse 'Italy * lambda * Infected'' failed, because of exception being raised:
E           SyntaxError: invalid syntax (<string>, line 1)

```
Similar errors were raised for 
BIOMD0000000971.xml
BIOMD0000000972.xml
BIOMD0000000983.xml



--------------------------------------------------------------------------------------------------------------------
There was one more kind of error which was possible raised due to the failed parsing of formula to ASTNode before extracting variables from the ```_get_variables()``` method. The error logs are as below:
```
num_children = astnode.getNumChildren()
E       AttributeError: 'NoneType' object has no attribute 'getNumChildren'

src/sbmlutils/report/mathml.py:108: AttributeError
```
The above error was raised in the following models:
BIOMD0000000760.xml
BIOMD0000000789.xml
BIOMD0000000791.xml

--------------------------------------------------------------------------------------------------------------------
A math related error was raised (```RisingFactorial```) for the model ```BIOMD0000000802.xml``` during its parsing. The error log generated is:
```
cls = RisingFactorial, args = (a1, y, R, A), options = {}, n = 4
temp = '%(name)s takes %(qual)s %(args)s argument%(plural)s (%(given)s given)'

    @cacheit
    def __new__(cls, *args, **options):
        # Handle calls like Function('f')
        if cls is Function:
            return UndefinedFunction(*args, **options)
    
        n = len(args)
        if n not in cls.nargs:
            # XXX: exception message must be in exactly this format to
            # make it work with NumPy's functions like vectorize(). See,
            # for example, https://github.com/numpy/numpy/issues/1697.
            # The ideal solution would be just to attach metadata to
            # the exception and change NumPy to take advantage of this.
            temp = ('%(name)s takes %(qual)s %(args)s '
                   'argument%(plural)s (%(given)s given)')
            raise TypeError(temp % {
                'name': cls,
                'qual': 'exactly' if len(cls.nargs) == 1 else 'at least',
                'args': min(cls.nargs),
                'plural': 's'*(min(cls.nargs) != 1),
>               'given': n})
E           TypeError: RisingFactorial takes exactly 2 arguments (4 given)

../../sb/lib/python3.7/site-packages/sympy/core/function.py:470: TypeError
```

--------------------------------------------------------------------------------------------------------------------

These were all the errors and associated models raised for all the models extracted from the URL. Although many of the models passed all the tests, the above few models failed in the "Latex" math type report generation. We can make fixes in the parsing of the specific formulae like the Piecewise functions after which these models should possible pass the tests.

Please guide me as to how we should proceed further with this issue. Also, if I should make any changes in the PR and model files uploaded, please let me know. Thanks!

matthiaskoenig commented on src/sbmlutils/test/__init__.py:
```diff
@@ -31,6 +32,993 @@
 REPRESSILATOR_SBML = MODELS_DIR / "repressilator" / "BIOMD0000000012_urn.xml"
 
+# Curated Biomodels
+BIOMD0000000001_SBML = MODELS_DIR / "biomodels_curated" / "BIOMD0000000001.xml"
```
This is extreme code duplication. Only store the path to the directory here and create the test files automatically! See comments below

matthiaskoenig commented on src/sbmlutils/test/report/test_curated_biomodels.py:
```diff
@@ -0,0 +1,2022 @@
+
+from sbmlutils.report import sbmlreport
+from sbmlutils.test import (
+    BIOMD0000000001_SBML,
```
Same here: extreme code duplication:
`from sbmlutils.test import BIOMODELS_PATH`

matthiaskoenig commented on src/sbmlutils/test/report/test_curated_biomodels.py:
```diff
@@ -0,0 +1,2022 @@
+)
+
+
+sbml_paths = [
```
Seriously this is a one liner?
```
sbml_paths = [BIOMODELS_PATH / f"BIOMD0000000{k:0>3}.xml" for k in range(1,988)]
```
Not sure the zero padding is correct, but you can figure this out, i.e. the `{k:0>3}`
Also update to compressed models

matthiaskoenig commented on src/sbmlutils/test/report/test_curated_biomodels.py:
```diff
@@ -0,0 +1,2022 @@
+    check_report_math_type(sbml_path=sbml_path, math_type="cmathml", tmp_path=tmp_path)
+
+
+def check_report_math_type(sbml_path: Path, math_type: str, tmp_path):
```
? please import this function. Again code duplication.

matthiaskoenig commented on src/sbmlutils/test/report/test_curated_biomodels.py:
```diff
@@ -0,0 +1,2022 @@
+]
+
+
+def sbml_paths_idfn(sbml_path):
```
import this function, code duplication

matthiaskoenig commented on src/sbmlutils/test/report/test_curated_biomodels.py:
```diff
@@ -0,0 +1,2022 @@
+    return sbml_path.name
+
+
+@pytest.mark.parametrize("sbml_path", sbml_paths, ids=sbml_paths_idfn)
```
remove this test. This is the same test as below with `latex`

matthiaskoenig reviewed (changes requested):
Some major changes. This is a lazy pull request. You have a ton of code duplication. What should be 20 lines are a few thousand.
Also use the compressed files and remove the uncompressed models from the history.
I will have a look at the failing models.

matthiaskoenig commented:
Initial feedback:
- please compress the files. You added ~120 MB to the repository, compressed the files are only around ~11MB. Compressed files can be read and should be used whenever possible. Make sure the uncompressed files are removed from the history so they do not bloat the repository.

matthiaskoenig commented:
I looked at the issue, but is distinct to the other problem before. The current issue is due to 
- symbol names in formulas which are restricted symbols such as `def` or `del`
- logical operations
I have to give a lot of lectures this week so have to see when I can fix this issue. Please update the pull request with the requested changes.

sankha555 commented:
@matthiaskoenig First of all, extremely sorry for the excessively duplicated code. It should have been intuitive to me that I should use compressed files for the models and not 100+ MB of individual files. The other instances of 1000 lines of code instead of 1-2 lines was also a really silly mistake on my side. I would make sure that these kind of trivial errors are not committed by me in future anymore, sorry again for the inconvenience caused.

I would significantly update the code and files to make the PR to its minimal size and maximum efficiency, and update the history. While this PR is open, I would continue studying the repository and other related resources for my preparation of the project. Whenever you can review the models according to your time and convenience, I would be eager to make any changes that are required according to your reviews. Thank you very much for the feedback and suggestions Sir.

matthiaskoenig commented:
Hi Sankha,

no worries. That is what code reviews are for.
Basically if you duplicate things or writing a lot of code to achieve small
things you should ask yourself if there is a simpler way to do this.
I assigned you to another smaller issue related to the report.
Next week we should start working on the design for the new report, i.e.
start sketching the work to do in the GSOC project.

Best Matthias


On Wed, Feb 17, 2021 at 5:59 AM Sankha Das <notifications@github.com> wrote:

> @matthiaskoenig <https://github.com/matthiaskoenig> First of all,
> extremely sorry for the excessively duplicated code. It should have been
> intuitive to me that I should use compressed files for the models and not
> 100+ MB of individual files. The other instances of 1000 lines of code
> instead of 1-2 lines was also a really silly mistake on my side. I would
> make sure that these kind of trivial errors are not committed by me in
> future anymore, sorry again for the inconvenience caused.
>
> I would significantly update the code and files to make the PR to its
> minimal size and maximum efficiency, and update the history. While this PR
> is open, I would continue studying the repository and other related
> resources for my preparation of the project. Whenever you can review the
> models according to your time and convenience, I would be eager to make any
> changes that are required according to your reviews. Thank you very much
> for the feedback and suggestions Sir.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/matthiaskoenig/sbmlutils/pull/207#issuecomment-780299360>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAG33OVND2LJ5CQ2BTA3OL3S7NEKTANCNFSM4XWNQ32Q>
> .
>


-- 
Matthias König, PhD.
Junior Group Leader LiSyM - Systems Medicine of the Liver
Humboldt Universität zu Berlin, Institute of Biology, Institute for
Theoretical Biology
  https://livermetabolism.com
konigmatt@googlemail.com
https://twitter.com/konigmatt
https://github.com/matthiaskoenig
Tel: +49 30 2093 98435

sankha555 commented:
Thank you very much for the reassurance and guidance Sir, I would keep these things in mind in future. I would start working on that issue (issue 144) as well as try to update this pull request. I had been a bit occupied with university coursework yesterday so couldn't do much work. I will start working again on both issues today afternoon onwards. 
I am eager to start working on the GSoC project for the new report and its proposal under your guidance. Looking forward to make great contributions in all the upcoming stages. 

Thanking You

Sincerely Yours,
Sankha Das

sankha555 commented:
@matthiaskoenig I have a query in the report generation for the compressed models. I have compressed all the 984 files into one .tar.gz archive and have kept it in the ```models``` directory. When I run the tests on the path for this archive (similar to the gz test for ```iAT_PLT_636.xml.gz``` archive), the test complete in a few seconds (which normally should take over 15 minutes, given the large no. of files). 

I think that the XML model files from within the archive are not being accessed by the test functions due to which all tests are being skipped for those files. Could you please help me out as to what could be the possible method I should use to run tests on the archive? I am using the following instances of tests:

In ```sbmlutils.test.__init__.py```
```
BIOMODELS_CURATED_PATH = MODELS_DIR / "biomodels_curated.tar.gz"
```

In ```test_curated_biomodels.py```
```
from sbmlutils.test import BIOMODELS_CURATED_PATH 

@pytest.mark.parametrize("sbml_path", [BIOMODELS_CURATED_PATH], ids=sbml_paths_idfn)
def test_report_gz(sbml_path, tmp_path):
    """Test report generation for GZ reports for curated biomodels."""
    sbmlreport.create_report(sbml_path=sbml_path, output_dir=tmp_path)

.
.
.
```
When I click on the .tar.gz archive in my system finder, it unpacks and returns a folder inside which the XML files are present. Could it be possible that due to this intermediate folder, that the tests are not able to run on the XML model files or is it an issue with the archive extension? 

Should I instead first unpack the archive using ```shutil.unpack_archive()``` and then run tests on each of the unpacked XML files? Please suggest me some hints about a correct direction to work on in order to run tests on the compressed models. I would make changes to the pull request and run the tests, and upload it for review. Thanks!

matthiaskoenig commented:
@Sankha Das <umasankhadas@gmail.com>
 Two solutions could work:
1. use individually compressed files; This will solve the issue and allow
to access the models individually (which is a nice feature for other tests).
(2. or write some pytest setup code, i.e. before the parametrized test is
executed the single file has to be extracted in a tmpdir which is used in
the tests.)

I clearly prefer solution 1. So you should just compress all the files
individually. You can easily do this via a shell script or adding a simple
python script. You can then use the one-liner to iterate over all the
compressed models in the parametrized test.
Other advantages of 1:
- it would allow simple update of models if new models are added (without a
10MB diff every time)
- compression should be faster using libsbml instead of python in a tmpdir

Best Matthias


On Thu, Feb 18, 2021 at 4:05 PM Sankha Das <notifications@github.com> wrote:

> @matthiaskoenig <https://github.com/matthiaskoenig> I have a query in the
> report generation for the compressed models. I have compressed all the 984
> files into one .tar.gz archive and have kept it in the models directory.
> When I run the tests on the path for this archive (similar to the gz test
> for iAT_PLT_636.xml.gz archive), the test complete in a few seconds
> (which normally should take over 15 minutes, given the large no. of files).
>
> I think that the XML model files from within the archive are not being
> accessed by the test functions due to which all tests are being skipped for
> those files. Could you please help me out as to what could be the possible
> method I should use to run tests on the archive? I am using the following
> instances of tests:
>
> In sbmlutils.test.__init__.py
>
> BIOMODELS_CURATED_PATH = MODELS_DIR / "biomodels_curated.tar.gz"
>
> In test_curated_biomodels.py
>
> from sbmlutils.test import BIOMODELS_CURATED_PATH
>
> @pytest.mark.parametrize("sbml_path", [BIOMODELS_CURATED_PATH], ids=sbml_paths_idfn)
> def test_report_gz(sbml_path, tmp_path):
>     """Test report generation for GZ reports for curated biomodels."""
>     sbmlreport.create_report(sbml_path=sbml_path, output_dir=tmp_path)
>
> ..
> .
> ..
>
> When I click on the .tar.gz archive in my system finder, it unpacks and
> returns a folder inside which the XML files are present. Could it be
> possible that due to this intermediate folder, that the tests are not able
> to run on the XML model files or is it an issue with the archive extension?
>
> Should I instead first unpack the archive using shutil.unpack_archive()
> and then run tests on each of the unpacked XML files? Please suggest me
> some hints about a correct direction to work on in order to run tests on
> the compressed models. I would make changes to the pull request and run the
> tests, and upload it for review. Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/matthiaskoenig/sbmlutils/pull/207#issuecomment-781406989>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAG33OWDI7ADXVA3ZJRUMSDS7UUCXANCNFSM4XWNQ32Q>
> .
>


-- 
Matthias König, PhD.
Junior Group Leader LiSyM - Systems Medicine of the Liver
Humboldt Universität zu Berlin, Institute of Biology, Institute for
Theoretical Biology
  https://livermetabolism.com
konigmatt@googlemail.com
https://twitter.com/konigmatt
https://github.com/matthiaskoenig
Tel: +49 30 2093 98435

sankha555 commented:
Thanks a lot for this explanation. I will start working on this approach and make the updates to pull request. The first approach seems more efficient to me too, I will make the script to compress the files individually and run the tests on them.

sankha555 commented:
@matthiaskoenig I have added the compressed models to the pull request (made a script using python gzip to compress all the models) and have also updated the tests and ```__init__.py``` files. Please take a look at the updates.

Regarding removing the history, I haven't made any changes because I am somewhat afraid that if it would fatally change the entire repository (this will actually be my first time removing history from a repository). Hence, I considered taking a confirmation from you for the upcoming steps that I should perform. I read this tutorial to remove the uncompressed files from the history : https://docs.github.com/en/github/authenticating-to-github/removing-sensitive-data-from-a-repository#using-filter-branch

Should I proceed with the steps in the above tutorial or is there any other workflow that I should follow? Also, will using these steps (```filter-branch```) remove the uncompressed files from all the branches (as it says) including develop? I am a bit worried that I don't remove all the files from develop too in the course of removing the uncompressed models from this branch. 

Could you please guide me if I am on the correct direction and if I should undertake any other steps? Also, please let me know if I have made any mistakes in the current updates, I would make the required changes to improve it. Thanks!

sankha555 commented:
I think I have made a mistake in updating the branch after rebasing it with the updated develop branch. The commit https://github.com/matthiaskoenig/sbmlutils/commit/dc26bd133766ec22acb1e7ef5c2f92c22d3c4252 that you had pushed to the branch, has been left out it in the recent push and is not showing in the commits. Is there any way that we can bring back the commit in the pull request Sir?

matthiaskoenig commented:
@Sankha Das <umasankhadas@gmail.com>
No worries. I will remove the uncompressed models using filter branch.
The commit you mention is just some reformating, so not important (that is
why it most likely got dropped).
I will wait with merging this branch until all the tests pass, i.e., after
I solved the issue with the symbols.

On Sat, Feb 20, 2021 at 9:48 AM Sankha Das <notifications@github.com> wrote:

> I think I have made a mistake in updating the branch after rebasing it
> with the updated develop branch. The commit dc26bd1
> <https://github.com/matthiaskoenig/sbmlutils/commit/dc26bd133766ec22acb1e7ef5c2f92c22d3c4252>
> that you had pushed to the branch, has been left out it in the recent push
> and is not showing in the commits. Is there any way that we can bring back
> the commit in the pull request Sir?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/matthiaskoenig/sbmlutils/pull/207#issuecomment-782589339>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAG33OSNULIXIRO3HNRSLP3S75ZOHANCNFSM4XWNQ32Q>
> .
>


-- 
Matthias König, PhD.
Junior Group Leader LiSyM - Systems Medicine of the Liver
Humboldt Universität zu Berlin, Institute of Biology, Institute for
Theoretical Biology
  https://livermetabolism.com
konigmatt@googlemail.com
https://twitter.com/konigmatt
https://github.com/matthiaskoenig
Tel: +49 30 2093 98435

sankha555 commented:
@matthiaskoenig Thanks a lot for the help, meanwhile I would also study how to do these operations like updating history  in detail so that I can do them myself in the future contributions. Please let me know in case I too should make any developments in the code for fixing the issue with the symbols, I would be eager to learn to do so. Thanks!
